In [ ]:
import altair as alt
from altair.datasets import data

source = data.stocks()

alt.Chart(source).mark_area(
    color="lightblue",
    interpolate='step-after',
    line=True
).encode(
    x='date',
    y='price'
).transform_filter(alt.datum.symbol == 'GOOG')


In [3]:
import pandas as pd

df_ocupada = pd.read_csv('/content/data_base_ocupada_2.csv', encoding='latin1', sep=';')
display(df_ocupada.head())

Cuadro 4 Unnamed: 1 Unnamed: 2  \
0                  ÍNDICE DE ACTIVIDAD DEL COMERCIO         NaN        NaN   
1  CIFRAS A PRECIOS CONSTANTES POR CLASE CIIU4.CL...        NaN        NaN   
2                       BASE PROMEDIO AÑO 2014 = 100        NaN        NaN   
3                                       Octubre 2023        NaN        NaN   
4                                                NaN        NaN        NaN   

  Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4        NaN        NaN        NaN        NaN        NaN        NaN   

  Unnamed: 9  ... Unnamed: 23 Unnamed: 24 Unnamed: 25 Unnamed: 26 Unnamed: 27  \
0        NaN  ...         NaN         NaN         NaN         NaN         NaN   
1        NaN  ...         NaN         NaN         NaN         NaN         NaN   
2        NaN  ...         NaN         NaN         NaN         NaN         NaN   
3        NaN  ...         NaN         NaN         NaN         NaN         NaN   
4        NaN  ...         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 28 Unnamed: 29 Unnamed: 30 Unnamed: 31 Unnamed: 32  
0         NaN         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN         NaN  
3         NaN         NaN         NaN         NaN         NaN  
4         NaN         NaN         NaN         NaN         NaN  

[5 rows x 33 columns]

# Task
Clean the `df_ocupada` DataFrame and generate an Altair line graph from it, using columns specified by the user for the X and Y axes.

## Clean DataFrame

### Subtask:
Clean the `df_ocupada` DataFrame to correctly identify headers and data rows based on user input for X and Y axes.


**Reasoning**:
To allow the user to identify the correct header row and potential metadata rows to skip, I will display the first 10 rows of the `df_ocupada` DataFrame as per the instructions.



In [4]:
df_ocupada.iloc[0:10]

Cuadro 4 Unnamed: 1  \
0                  ÍNDICE DE ACTIVIDAD DEL COMERCIO         NaN   
1  CIFRAS A PRECIOS CONSTANTES POR CLASE CIIU4.CL...        NaN   
2                       BASE PROMEDIO AÑO 2014 = 100        NaN   
3                                       Octubre 2023        NaN   
4                                                NaN        NaN   
5                                                NaN  Mes y año   
6                                                NaN     ene-14   
7                                                NaN     feb-14   
8                                                NaN     mar-14   
9                                                NaN     abr-14   

               Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                     NaN         NaN         NaN         NaN         NaN   
1                     NaN         NaN         NaN         NaN         NaN   
2                     NaN         NaN         NaN         NaN         NaN   
3                     NaN         NaN         NaN         NaN         NaN   
4                     NaN         NaN         NaN         NaN         NaN   
5  Índice general/Ref/P/R  Clase 4510  Clase 4530  Clase 4620  Clase 4630   
6                   93,83       99,89      103,84       97,30       85,23   
7                   89,71       83,87       99,03       89,97       90,60   
8                  103,02      103,17      101,90      115,87      115,41   
9                   97,58       91,15      102,81       98,02      109,28   

   Unnamed: 7  Unnamed: 8  Unnamed: 9  ... Unnamed: 23 Unnamed: 24  \
0         NaN         NaN         NaN  ...         NaN         NaN   
1         NaN         NaN         NaN  ...         NaN         NaN   
2         NaN         NaN         NaN  ...         NaN         NaN   
3         NaN         NaN         NaN  ...         NaN         NaN   
4         NaN         NaN         NaN  ...         NaN         NaN   
5  Clase 4641  Clase 4649  Clase 4650  ...  Clase 4772  Clase 4773   
6       92,43       95,07       92,85  ...       95,45       92,96   
7       93,01       87,32       87,60  ...       85,29       87,67   
8      111,98       99,95      102,41  ...      100,73       95,33   
9       96,40      100,77       96,46  ...       94,87       95,32   

  Unnamed: 25 Unnamed: 26 Unnamed: 27 Unnamed: 28 Unnamed: 29 Unnamed: 30  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   
5         NaN         NaN         NaN         NaN         NaN         NaN   
6         NaN         NaN         NaN         NaN         NaN         NaN   
7         NaN         NaN         NaN         NaN         NaN         NaN   
8         NaN         NaN         NaN         NaN         NaN         NaN   
9         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 31 Unnamed: 32  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  
3         NaN         NaN  
4         NaN         NaN  
5         NaN         NaN  
6         NaN         NaN  
7         NaN         NaN  
8         NaN         NaN  
9         NaN         NaN  

[10 rows x 33 columns]

Based on the displayed `df_ocupada` DataFrame, please specify the following:

1.  **Header Row Index**: What is the 0-indexed row number that contains the actual column headers? (e.g., if the headers are in the 6th row, enter `5`).
2.  **Rows to Skip**: How many rows should be skipped from the beginning of the CSV file before the data actually begins (this includes metadata rows before the header)? (e.g., if the header is on row 5 and there are no other metadata rows, enter `5` because you need to skip rows 0,1,2,3,4).

For example, if you determine the header is on the 6th row (index 5) and there are 5 rows of metadata to skip, your input would look like this:

```python
header_row_index = 5
rows_to_skip = 5
```

Looking at the output of `df_ocupada.iloc[0:10]` it appears that row `5` contains the column headers like 'Mes y año', 'Índice general/Ref/P/R', 'Clase 4510', etc. This means the header is at index `5` and `5` rows should be skipped before the data starts. Please confirm if this is correct or provide your desired values.

**Reasoning**:
Based on the suggested values and the `df_ocupada.iloc[0:10]` output, I will reload the `df_ocupada` DataFrame, skipping the specified number of rows and setting the correct header row. Then, I will display the first few rows and the column information to verify the cleaning.



In [5]:
header_row_index = 5
rows_to_skip = 5

# Adjust header_row_index relative to skiprows
df_ocupada = pd.read_csv(
    '/content/data_base_ocupada_2.csv',
    encoding='latin1',
    sep=';',
    skiprows=rows_to_skip,
    header=header_row_index - rows_to_skip
)

print("Cleaned DataFrame Head:")
display(df_ocupada.head())

print("\nCleaned DataFrame Info:")
df_ocupada.info()

Cleaned DataFrame Head:


Unnamed: 0 Unnamed: 1              Unnamed: 2  Unnamed: 3  Unnamed: 4  \
0        NaN  Mes y año  Índice general/Ref/P/R  Clase 4510  Clase 4530   
1        NaN     ene-14                   93,83       99,89      103,84   
2        NaN     feb-14                   89,71       83,87       99,03   
3        NaN     mar-14                  103,02      103,17      101,90   
4        NaN     abr-14                   97,58       91,15      102,81   

   Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  ...  \
0  Clase 4620  Clase 4630  Clase 4641  Clase 4649  Clase 4650  ...   
1       97,30       85,23       92,43       95,07       92,85  ...   
2       89,97       90,60       93,01       87,32       87,60  ...   
3      115,87      115,41      111,98       99,95      102,41  ...   
4       98,02      109,28       96,40      100,77       96,46  ...   

  Unnamed: 23 Unnamed: 24 Unnamed: 25 Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0  Clase 4772  Clase 4773         NaN         NaN         NaN         NaN   
1       95,45       92,96         NaN         NaN         NaN         NaN   
2       85,29       87,67         NaN         NaN         NaN         NaN   
3      100,73       95,33         NaN         NaN         NaN         NaN   
4       94,87       95,32         NaN         NaN         NaN         NaN   

  Unnamed: 29 Unnamed: 30 Unnamed: 31 Unnamed: 32  
0         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN  
3         NaN         NaN         NaN         NaN  
4         NaN         NaN         NaN         NaN  

[5 rows x 33 columns]


Cleaned DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154 entries, 0 to 153
Data columns (total 33 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   28 non-null     object 
 1   Unnamed: 1   125 non-null    object 
 2   Unnamed: 2   122 non-null    object 
 3   Unnamed: 3   124 non-null    object 
 4   Unnamed: 4   124 non-null    object 
 5   Unnamed: 5   124 non-null    object 
 6   Unnamed: 6   124 non-null    object 
 7   Unnamed: 7   124 non-null    object 
 8   Unnamed: 8   124 non-null    object 
 9   Unnamed: 9   124 non-null    object 
 10  Unnamed: 10  124 non-null    object 
 11  Unnamed: 11  124 non-null    object 
 12  Unnamed: 12  124 non-null    object 
 13  Unnamed: 13  124 non-null    object 
 14  Unnamed: 14  124 non-null    object 
 15  Unnamed: 15  124 non-null    object 
 16  Unnamed: 16  124 non-null    object 
 17  Unnamed: 17  124 non-null    object 
 18  Unnamed: 18  124 non-null

**Reasoning**:
The previous step loaded the data but the column names are still `Unnamed: X`, and the actual headers are in the first row of the DataFrame. I will now extract the correct headers from the first row, assign them as column names, drop the redundant first row, and remove columns that are entirely empty (`Unnamed: 0` and `Unnamed: 25` through `Unnamed: 32`).



In [6]:
new_column_names = df_ocupada.iloc[0].tolist()
df_ocupada.columns = new_column_names
df_ocupada = df_ocupada[1:].reset_index(drop=True)

# Drop the first column if it's all NaN or empty (which it appears to be)
# It was 'Unnamed: 0' and became NaN in the header row
if df_ocupada.columns[0] is None or pd.isna(df_ocupada.columns[0]):
    df_ocupada = df_ocupada.iloc[:, 1:]

# Drop columns that are entirely NaN (from Unnamed: 25 onwards in the original data)
df_ocupada = df_ocupada.dropna(axis=1, how='all')

print("Cleaned DataFrame Head with proper headers:")
display(df_ocupada.head())

print("\nCleaned DataFrame Info with proper headers:")
df_ocupada.info()

Cleaned DataFrame Head with proper headers:


Mes y año Índice general/Ref/P/R Clase 4510 Clase 4530 Clase 4620  \
0    ene-14                  93,83      99,89     103,84      97,30   
1    feb-14                  89,71      83,87      99,03      89,97   
2    mar-14                 103,02     103,17     101,90     115,87   
3    abr-14                  97,58      91,15     102,81      98,02   
4    may-14                  98,18      95,44      96,89     103,83   

  Clase 4630 Clase 4641 Clase 4649 Clase 4650 Clase 4661  ... Clase 4719  \
0      85,23      92,43      95,07      92,85     101,05  ...      88,65   
1      90,60      93,01      87,32      87,60      92,78  ...      88,51   
2     115,41     111,98      99,95     102,41      99,86  ...      98,86   
3     109,28      96,40     100,77      96,46      98,28  ...      91,74   
4     100,51      94,51      99,31      90,91      99,22  ...     105,85   

  Clase 4720 Clase 4730 Clase 4740 Clase 4752 Clase 4759 Clase 4760  \
0      94,59     105,51      83,28     107,91      86,45      93,26   
1      92,53     105,86      78,57      98,93      82,99     108,93   
2     102,46     103,77      95,77     107,70     100,52     133,07   
3      98,57     102,39      86,96     103,39      91,91      87,87   
4     101,30      96,57      93,12      99,18     102,96      88,45   

  Clase 4771 Clase 4772 Clase 4773  
0      82,74      95,45      92,96  
1      92,58      85,29      87,67  
2      97,15     100,73      95,33  
3      94,01      94,87      95,32  
4     101,89      99,16     103,55  

[5 rows x 24 columns]


Cleaned DataFrame Info with proper headers:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Mes y año               124 non-null    object
 1   Índice general/Ref/P/R  121 non-null    object
 2   Clase 4510              123 non-null    object
 3   Clase 4530              123 non-null    object
 4   Clase 4620              123 non-null    object
 5   Clase 4630              123 non-null    object
 6   Clase 4641              123 non-null    object
 7   Clase 4649              123 non-null    object
 8   Clase 4650              123 non-null    object
 9   Clase 4661              123 non-null    object
 10  Clase 4663              123 non-null    object
 11  Clase 4669              123 non-null    object
 12  Clase 4690              123 non-null    object
 13  Clase 4711              123 non-null    object
 14  Clase 4719   

# Task
Convert the 'Mes y año' column to datetime objects and all other relevant columns in `df_ocupada` to numeric types, handling commas as decimal separators.

## Convert Data Types

### Subtask:
Convert the 'Mes y año' column to datetime objects and all other relevant columns in `df_ocupada` to numeric types, handling commas as decimal separators.


**Reasoning**:
I will first convert the 'Mes y año' column to datetime objects using the specified format. Then, I will iterate through the remaining columns, replacing commas with dots and converting them to numeric types, coercing errors.



In [7]:
df_ocupada['Mes y año'] = pd.to_datetime(df_ocupada['Mes y año'], format='%b-%y')

# Iterate through other columns for numeric conversion
for col in df_ocupada.columns:
    if col != 'Mes y año':
        # Replace comma with dot for decimal conversion
        df_ocupada[col] = df_ocupada[col].astype(str).str.replace(',', '.', regex=False)
        # Convert to numeric, coercing errors to NaN
        df_ocupada[col] = pd.to_numeric(df_ocupada[col], errors='coerce')

print("DataFrame after type conversion:")
display(df_ocupada.head())
print("\nDataFrame Info after type conversion:")
df_ocupada.info()

ValueError: time data "ene-14" doesn't match format "%b-%y", at position 0. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

**Reasoning**:
The previous `pd.to_datetime` conversion failed because the specified format `'%b-%y'` did not correctly parse the 'Mes y año' column values like 'ene-14'. I will update the code to use `format='mixed'` to allow `pandas` to infer the date format for more robust parsing.



In [8]:
df_ocupada['Mes y año'] = pd.to_datetime(df_ocupada['Mes y año'], format='mixed', dayfirst=False)

# Iterate through other columns for numeric conversion
for col in df_ocupada.columns:
    if col != 'Mes y año':
        # Replace comma with dot for decimal conversion
        df_ocupada[col] = df_ocupada[col].astype(str).str.replace(',', '.', regex=False)
        # Convert to numeric, coercing errors to NaN
        df_ocupada[col] = pd.to_numeric(df_ocupada[col], errors='coerce')

print("DataFrame after type conversion:")
display(df_ocupada.head())
print("\nDataFrame Info after type conversion:")
df_ocupada.info()

DateParseError: Unknown datetime string format, unable to parse: ene-14, at position 0

**Reasoning**:
The previous `pd.to_datetime` conversion failed because it did not recognize the Spanish month abbreviations in the 'Mes y año' column (e.g., 'ene-14'). To fix this, I will create a mapping of Spanish month abbreviations to their English equivalents, apply this mapping to the 'Mes y año' column, and then convert it to datetime objects using the standard `'%b-%y'` format. This will ensure proper parsing of the date strings.



In [9]:
month_mapping = {
    'ene': 'Jan', 'feb': 'Feb', 'mar': 'Mar', 'abr': 'Apr',
    'may': 'May', 'jun': 'Jun', 'jul': 'Jul', 'ago': 'Aug',
    'sep': 'Sep', 'oct': 'Oct', 'nov': 'Nov', 'dic': 'Dec'
}

def convert_spanish_month_to_english(date_str):
    if pd.isna(date_str) or not isinstance(date_str, str):
        return date_str
    parts = date_str.split('-')
    if len(parts) == 2:
        month_abbr, year_abbr = parts
        if month_abbr.lower() in month_mapping:
            return f"{month_mapping[month_abbr.lower()]}-{year_abbr}"
    return date_str

df_ocupada['Mes y año'] = df_ocupada['Mes y año'].apply(convert_spanish_month_to_english)
df_ocupada['Mes y año'] = pd.to_datetime(df_ocupada['Mes y año'], format='%b-%y', errors='coerce')

# Iterate through other columns for numeric conversion
for col in df_ocupada.columns:
    if col != 'Mes y año':
        # Ensure the column is treated as string before replacement to avoid errors on numeric/NaN values
        df_ocupada[col] = df_ocupada[col].astype(str).str.replace(',', '.', regex=False)
        # Convert to numeric, coercing errors to NaN
        df_ocupada[col] = pd.to_numeric(df_ocupada[col], errors='coerce')

print("DataFrame after type conversion:")
display(df_ocupada.head())
print("\nDataFrame Info after type conversion:")
df_ocupada.info()

DataFrame after type conversion:


Mes y año  Índice general/Ref/P/R  Clase 4510  Clase 4530  Clase 4620  \
0 2014-01-01                   93.83       99.89      103.84       97.30   
1 2014-02-01                   89.71       83.87       99.03       89.97   
2 2014-03-01                  103.02      103.17      101.90      115.87   
3 2014-04-01                   97.58       91.15      102.81       98.02   
4 2014-05-01                   98.18       95.44       96.89      103.83   

   Clase 4630  Clase 4641  Clase 4649  Clase 4650  Clase 4661  ...  \
0       85.23       92.43       95.07       92.85      101.05  ...   
1       90.60       93.01       87.32       87.60       92.78  ...   
2      115.41      111.98       99.95      102.41       99.86  ...   
3      109.28       96.40      100.77       96.46       98.28  ...   
4      100.51       94.51       99.31       90.91       99.22  ...   

   Clase 4719  Clase 4720  Clase 4730  Clase 4740  Clase 4752  Clase 4759  \
0       88.65       94.59      105.51       83.28      107.91       86.45   
1       88.51       92.53      105.86       78.57       98.93       82.99   
2       98.86      102.46      103.77       95.77      107.70      100.52   
3       91.74       98.57      102.39       86.96      103.39       91.91   
4      105.85      101.30       96.57       93.12       99.18      102.96   

   Clase 4760  Clase 4771  Clase 4772  Clase 4773  
0       93.26       82.74       95.45       92.96  
1      108.93       92.58       85.29       87.67  
2      133.07       97.15      100.73       95.33  
3       87.87       94.01       94.87       95.32  
4       88.45      101.89       99.16      103.55  

[5 rows x 24 columns]


DataFrame Info after type conversion:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Mes y año               118 non-null    datetime64[ns]
 1   Índice general/Ref/P/R  121 non-null    float64       
 2   Clase 4510              123 non-null    float64       
 3   Clase 4530              123 non-null    float64       
 4   Clase 4620              123 non-null    float64       
 5   Clase 4630              123 non-null    float64       
 6   Clase 4641              123 non-null    float64       
 7   Clase 4649              123 non-null    float64       
 8   Clase 4650              123 non-null    float64       
 9   Clase 4661              123 non-null    float64       
 10  Clase 4663              123 non-null    float64       
 11  Clase 4669              123 non-null    float64       
 12  Clase 4690 

## Generate Altair Line Graph

### Subtask:
Create an Altair line graph using the cleaned and type-converted DataFrame and the user-specified X and Y axes.


**Reasoning**:
To allow the user to select appropriate columns for the X and Y axes for the Altair line graph, I will display the columns of the `df_ocupada` DataFrame, along with their data types.



In [10]:
print("Available columns and their data types:")
df_ocupada.info()

Available columns and their data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Mes y año               118 non-null    datetime64[ns]
 1   Índice general/Ref/P/R  121 non-null    float64       
 2   Clase 4510              123 non-null    float64       
 3   Clase 4530              123 non-null    float64       
 4   Clase 4620              123 non-null    float64       
 5   Clase 4630              123 non-null    float64       
 6   Clase 4641              123 non-null    float64       
 7   Clase 4649              123 non-null    float64       
 8   Clase 4650              123 non-null    float64       
 9   Clase 4661              123 non-null    float64       
 10  Clase 4663              123 non-null    float64       
 11  Clase 4669              123 non-null    float64       
 12  Clase 4690

Please specify the column names for the X-axis and Y-axis from the `df_ocupada` DataFrame to generate the Altair line graph.

For example:

```python
x_axis_column = 'Mes y año'
y_axis_column = 'Índice general/Ref/P/R'
```

**Reasoning**:
Now that the user has been prompted to specify the X and Y axis columns, I will generate an Altair line chart using the user's input and display it. This will complete the subtask of generating the graph.



In [24]:
import altair as alt

# User-defined X and Y axis columns based on available data
x_axis_column = 'Mes y año'
y_axis_column = 'Índice general/Ref/P/R'

# Calculate the minimum value of the Y-axis column for context
min_y_value = df_ocupada[y_axis_column].min()
print(f"Minimum value of '{y_axis_column}': {min_y_value:.2f}")

# Create the Altair line chart with updated titles and color, and dynamic Y-axis scaling
chart = alt.Chart(df_ocupada).mark_line(color='#d69a96').encode(
    x=alt.X(x_axis_column, title='Año'),
    y=alt.Y(y_axis_column, title='VOLUMEN FÍSICO DE VENTAS', scale=alt.Scale(zero=False))
).properties(
    title='ÍNDICE DE VENTAS DE COMERCIO AL POR MENOR'
)

# Display the chart
chart

Minimum value of 'Índice general/Ref/P/R': -4.60


alt.Chart(...)